compiler/rustc_lint/src/mutable_transmutes.rs

```diff
@@ -0,0 +1,721 @@
+            idx += 1;
+            continue;
+        }
+
+        let span = debug_span!("on_overlapping", ?ty, ?idx, ?fields_breadcrumbs);
+        let _guard = span.enter();
+        let result = on_overlapping(ty, idx, &fields_breadcrumbs);
+        debug!(?result, "on_overlapping result");
+        match result {
+            CheckOverlapping::Yes => {}
+            CheckOverlapping::No => return,
+        };
+        idx += 1;
+    }
+
+    // Finally, descend to every field and check there too.
+    match ty.kind() {
+        &ty::Adt(adt_def, substs) if adt_def.is_struct() => {
+            for (field_idx, field_offset, field_ty) in
+                struct_fields(cx, layout.layout, adt_def, substs)
+            {
+                fields_breadcrumbs.push_with(field_idx, |fields_breadcrumbs| {
+                    check_overlapping(
+                        cx,
+                        check_overlap_with,
+                        field_ty,
+                        offset + field_offset,
+                        on_overlapping,
+                        fields_breadcrumbs,
+                    )
+                });
+            }
+        }
+
+        &ty::Tuple(tys) => {
+            for (field_idx, field_offset, field_ty) in tuple_fields(layout.layout, tys) {
+                fields_breadcrumbs.push_with(field_idx, |fields_breadcrumbs| {
+                    check_overlapping(
+                        cx,
+                        check_overlap_with,
+                        field_ty,
+                        offset + field_offset,
+                        on_overlapping,
+                        fields_breadcrumbs,
+                    )
+                });
+            }
+        }
+
+        &ty::Array(ty, _) if layout.size != Size::ZERO => {
+            fields_breadcrumbs.push_with(FieldIdx::from_u32(0), |fields_breadcrumbs| {
+                check_overlapping(
+                    cx,
+                    check_overlap_with,
+                    ty,
+                    offset,
+                    on_overlapping,
+                    fields_breadcrumbs,
+                )
+            });
+        }
+
+        _ => {}
+    }
+}
+
+#[derive(Debug)]
+struct Ref<'tcx> {
+    start_offset: Size,
+    end_offset: Size,
+    ty: Ty<'tcx>,
+    mutability: Mutability,
+    fields_breadcrumbs: FieldsBreadcrumbs,
+}
+
+impl Field for Ref<'_> {
+    fn start_offset(&self) -> Size {
+        self.start_offset
+    }
+
+    fn end_offset(&self) -> Size {
+        self.end_offset
+    }
+}
+
+#[derive(Debug)]
+struct UnsafeCellField {
+    start_offset: Size,
+    end_offset: Size,
+    fields_breadcrumbs: FieldsBreadcrumbs,
+}
+
+impl Field for UnsafeCellField {
+    fn start_offset(&self) -> Size {
+        self.start_offset
+    }
+
+    fn end_offset(&self) -> Size {
+        self.end_offset
+    }
+}
+
+/// The parameters to `report_error` are: breadcrumbs to src's UnsafeCell, breadcrumbs to dst's UnsafeCell.
+fn check_unsafe_cells<'tcx>(
```

Comment:
I wonder why the `is_freeze` query doesn't work for this?

Reply:
It can work for finding `&` to `&UnsafeCell` casts, but not for finding `&` to `&mut` or `&` to `&UnsafeCell` inside fields. So we need to have this anyway, given that it's just natural to use it for unsafe cell too.